pq-crypto/s2n_pq.c

```diff
@@ -82,19 +95,115 @@ bool s2n_cpu_supports_sikep434r2_asm() {
 #endif /* defined(S2N_SIKEP434R2_ASM) */
 }
 
+bool s2n_cpu_supports_bike_r3_avx2() {
+#if defined(S2N_BIKE_R3_AVX2)
+    uint32_t eax, ebx, ecx, edx;
+    if (!s2n_get_cpuid_count(EXTENDED_FEATURES_LEAF, EXTENDED_FEATURES_SUBLEAF_ZERO, &eax, &ebx, &ecx, &edx)) {
+        return false;
+    }
+    return ((ebx & EBX_BIT_AVX2) != 0);
+#else
+    return false;
+#endif
+}
+
+bool s2n_cpu_supports_bike_r3_avx512() {
+#if defined(S2N_BIKE_R3_AVX512)
+    uint32_t eax, ebx, ecx, edx;
+    if (!s2n_get_cpuid_count(EXTENDED_FEATURES_LEAF, EXTENDED_FEATURES_SUBLEAF_ZERO, &eax, &ebx, &ecx, &edx)) {
+        return false;
+    }
+    return ((ebx & EBX_BIT_AVX512) != 0);
+#else
+    return false;
+#endif
+}
+
+bool s2n_cpu_supports_bike_r3_pclmul() {
+#if defined(S2N_BIKE_R3_PCLMUL)
+    uint32_t eax, ebx, ecx, edx;
+    if (!s2n_get_cpuid_count(PROCESSOR_INFO_AND_FEATURES, EXTENDED_FEATURES_SUBLEAF_ZERO, &eax, &ebx, &ecx, &edx)) {
+        return false;
+    }
+    return ((ecx & ECX_BIT_PCLMUL) != 0);
+#else
+    return false;
+#endif
+}
+
+bool s2n_cpu_supports_bike_r3_vpclmul() {
+#if defined(S2N_BIKE_R3_AVX512)
+    uint32_t eax, ebx, ecx, edx;
+    if (!s2n_get_cpuid_count(EXTENDED_FEATURES_LEAF, EXTENDED_FEATURES_SUBLEAF_ZERO, &eax, &ebx, &ecx, &edx)) {
+        return false;
+    }
+    return ((ecx & ECX_BIT_VPCLMUL) != 0);
+#else
+    return false;
+#endif
+}
+
+S2N_RESULT s2n_bike_r3_x86_64_opt_init() {
+    bike_r3_avx2_enabled    = s2n_cpu_supports_bike_r3_avx2();
+    bike_r3_avx512_enabled  = s2n_cpu_supports_bike_r3_avx512();
+    bike_r3_pclmul_enabled  = s2n_cpu_supports_bike_r3_pclmul();
+    bike_r3_vpclmul_enabled = s2n_cpu_supports_bike_r3_vpclmul();
+    return S2N_RESULT_OK;
+}
```

Comment:
I think there's a little conflation between the `opt_init` and `try_enable` functions. I think all of the `try_enable_bike_r3*` functions as written below are correct, and this function should just wrap them all up for convenience. Something like:

```
S2N_RESULT s2n_bike_r3_x86_64_opt_init() {
    RESULT_ENSURE_OK(s2n_try_enable_bike_r3_opt_pclmul());
    RESULT_ENSURE_OK(s2n_try_enable_bike_r3_opt_avx2());
    RESULT_ENSURE_OK(s2n_try_enable_bike_r3_opt_avx512());
    RESULT_ENSURE_OK(s2n_try_enable_bike_r3_opt_vpclmul());
    return S2N_RESULT_OK;
}
```

And then we can define this function only once (outside of the CPUID ifdefs), down where all the `try_enable` functions are defined.

Reply:
done